drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,2349 @@
+{
+	const struct dw_i3c_config *config = dev->config;
+	struct dw_i3c_data *data = dev->data;
+	uint32_t status;
+	uint32_t present_state;
+
+	status = sys_read32(config->regs + INTR_STATUS);
+	if (status & (INTR_TRANSFER_ERR_STAT | INTR_RESP_READY_STAT)) {
+		dw_i3c_end_xfer(dev);
+
+		if (status & INTR_TRANSFER_ERR_STAT) {
+			sys_write32(INTR_TRANSFER_ERR_STAT, config->regs + INTR_STATUS);
+		}
+	}
+
+	if (status & INTR_IBI_THLD_STAT) {
+#ifdef CONFIG_I3C_USE_IBI
+		ibis_handle(dev);
+#endif /* CONFIG_I3C_USE_IBI */
+	}
+
+	/* target mode related interrupts */
+	present_state = sys_read32(config->regs + PRESENT_STATE);
+	if (!(present_state & PRESENT_STATE_CURRENT_MASTER)) {
+		const struct i3c_target_callbacks *target_cb = data->target_config->callbacks;
+
+		/* Data has been written to the rx fifo by a controller while in target mode*/
+		if (status & INTR_RX_THLD_STAT) {
+			/* while the rx fifo buffer is not empty */
+			while (DATA_BUFFER_STATUS_LEVEL_RX(
+				       sys_read32(config->regs + DATA_BUFFER_STATUS_LEVEL)) > 0) {
+				uint8_t rx_data;
+				/*
+				 * TODO: verify how byte size (lol) data is written to a dword size
+				 * buffer over i3c in target mode
+				 */
+				read_rx_fifo(dev, &rx_data, sizeof(rx_data));
+				/* call function pointer for write */
+				if (target_cb != NULL && target_cb->write_received_cb != NULL) {
+					target_cb->write_received_cb(data->target_config, rx_data);
+				}
+			}
+			sys_write32(INTR_RX_THLD_STAT, config->regs + INTR_STATUS);
+		}
+		/* Read Requested when the CMDQ is empty*/
+		if (status & INTR_READ_REQ_RECV_STAT) {
+			if (target_cb != NULL && target_cb->write_received_cb != NULL) {
+				/* TODO: how should the first byte be handled? */
+				target_cb->read_requested_cb(data->target_config, NULL);
+			}
+			sys_write32(INTR_READ_REQ_RECV_STAT, config->regs + INTR_STATUS);
+		}
+#ifdef CONFIG_I3C_USE_IBI
+		/* IBI TIR request register is addressed and status is updated*/
+		if (status & INTR_IBI_UPDATED_STAT) {
+			k_sem_give(&data->ibi_sts_sem);
+			sys_write32(INTR_IBI_UPDATED_STAT, config->regs + INTR_STATUS);
+		}
+		/* DA has been assigned, could happen after a IBI HJ request */
+		if (status & INTR_DYN_ADDR_ASSGN_STAT) {
+			/* TODO: handle IBI HJ with semaphore */
+			sys_write32(INTR_DYN_ADDR_ASSGN_STAT, config->regs + INTR_STATUS);
+		}
+#endif /* CONFIG_I3C_USE_IBI */
+	}
+
+	return 0;
+}
+
+static int init_scl_timing(const struct device *dev)
+{
+	const struct dw_i3c_config *config = dev->config;
+	struct dw_i3c_data *data = dev->data;
+	uint32_t scl_timing, hcnt, lcnt, core_rate;
+
+	if (clock_control_get_rate(config->clock, NULL, &core_rate) != 0) {
+		LOG_ERR("%s: get clock rate failed", dev->name);
+		return -EINVAL;
+	}
+
+	/* I3C_OD */
+	hcnt = DIV_ROUND_UP(config->od_thigh_max_ns * core_rate, I3C_PERIOD_NS) - 1;
+	hcnt = CLAMP(hcnt, SCL_I3C_TIMING_CNT_MIN, SCL_I3C_TIMING_CNT_MAX);
+
+	lcnt = DIV_ROUND_UP(config->od_tlow_min_ns * core_rate, I3C_PERIOD_NS);
```

Comment:
Will it not overflow when:
1. core_rate is 100MHz
2. config->od_tlow_min_ns = 200

Reply:
this was something we caught internally a while ago, I just haven't push a new commit fixing that issue... but the solution was to just cast `core_rate` to a `uint64_t`